### DISCUSSION TITLE
Divergence approaching steady-state

### DISCUSSION THREAD

**Discussion by Khaksarkn** [idaholab/moose#26094] [Q&A General] [answered]
Hi,

I have written a custom MOOSE code to solve a THM problem. It works well while it is transient.
I use -lu preconditioning and manual or automatic scaling. 

However, when the solution is approaching a steady state, it starts to diverge. Using automatic scaling is more effective and the solution goes further and sometimes actually reaches a steady state. But I assume when it reaches steady conditions, it is supposed to go on, not diverge. 

I would appreciate it if you could help me figure out this issue.
Thank you.

**Comment by GiudGiud** [comment]
Hello

Can you please paste the console log here?
Could you please add `Executioner/verbose=true` or some other option to get the scaling factors to be output to console?
Automatic scaling could get tricky next to steady state. Make sure to NOT set `compute_scaling_once = false`

Guillaume

**Reply by Khaksarkn** [reply]
I also need to mention this. When I set the BCs such that the model is already at a steady state, it struggles to proceed and diverges (which is the same as when it reaches a steady state from a transient procedure).

Thanks, This is not set as false.
compute_scaling_once = false

This is a part of the console:

Time Step 65, time = 49950
          old time = 49800
                dt = 150
            old dt = 300

 0 Nonlinear |R| = 1.848899e-04
 1 Nonlinear |R| = 1.804325e-12
 Solve Converged!

Outlier Variable Residual Norms:
  porePressure: 1.622165e-12

Time Step 66, time = 50250
          old time = 49950
                dt = 300
            old dt = 150

 0 Nonlinear |R| = 1.844882e-04
 1 Nonlinear |R| = 1.888366e-12
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 1
 Solve Did NOT Converge!
Aborting as solve did not converge

**Reply by GiudGiud** [reply]
it must be:
`compute_scaling_once = true`
false will cause issues

**Reply by Khaksarkn** [reply]
Yes, it is set as "true". I double-checked.

**Reply by GiudGiud** [reply]
1 Nonlinear |R| = 1.804325e-12
is converged

at this point you just need to adapt the tolerance to avoid trying to over-converge beyond what the numerical errors will allow

**Reply by Khaksarkn** [reply]
Thank you for the input. 

Yes, the 1.888366e-12 value when it says it has not converged also looks like it has converged.

Can you please tell me how I should do that? 

When I used to solve simple diffusion problems, even if I had already steady state BC, it just carried on with 1 nonlinear iteration untill end_time occured.

**Reply by GiudGiud** [reply, marked as answer]
nl_abs_tol = 2e-12 will let it accept a solution with an error that is below that. You would be fine for this time step.

**Reply by Khaksarkn** [reply]
I just have one other question, 
How can I put a distributed load on my mesh?
I want a 1000 N/m load on a 2D model on one of the boundaries. Is it through NeumanBC for disp_y (y component of the displacement) and setting the value to the load? It seems a bit strange looking at the formulation of the NeumannBC.

Thanks.

**Reply by GiudGiud** [reply]
I think the Pressure action can be used too.
https://mooseframework.inl.gov/syntax/BCs/Pressure/index.html
NeumannBC should be fine. Just make sure the load is normalized to the surface, I see N/m I think it should be N/m2?

**Reply by Khaksarkn** [reply]
Thank you Guillaume for the help. I also needed to set `-snes_atol` to resolve the issue. Now I can reach the steady-state solution.

Is the pressure kernel a part of the `TensorMechanics` Module? I activated TensorMechanics in my app but I'm still getting errors:

[load]
    type = Pressure
    variable = disp_y
    boundary = loadBC
    factor = -1
    displacements = 'disp_x disp_y'
  []
*** ERROR ***
A 'Pressure' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

**Reply by GiudGiud** [reply]
It is. 
You ll have to double check the syntax maybe?
I think it's like this `[Pressure]`

**Reply by GiudGiud** [reply]
```

[BCs]
  [./no_x]
    type = DirichletBC
    variable = disp_x
    boundary = 4
    value = 0.0
  [../]
  [./no_y]
    type = DirichletBC
    variable = disp_y
    boundary = 5
    value = 0.0
  [../]
  [./no_z]
    type = DirichletBC
    variable = disp_z
    boundary = 6
    value = 0.0
  [../]
  [./Pressure]
    [./Side1]
      boundary = 1
      function = rampConstant
      displacements = 'disp_x disp_y disp_z'
    [../]
```
nested twice
